packages/astro/src/core/render/core.ts

```diff
@@ -101,26 +107,50 @@ export async function tryRenderPage<MiddlewareReturnType = Response>(
 		adapterName: env.adapterName,
 	});
 
-	if (onRequest) {
-		return await callMiddleware<Response>(
-			env.logging,
-			onRequest as MiddlewareResponseHandler,
-			apiContext,
-			() => {
-				return renderPage({
+	switch (routeType) {
+		case 'page':
+		case 'redirect': {
+			if (onRequest) {
+				return await callMiddleware<Response>(
+					env.logging,
+					onRequest as MiddlewareResponseHandler,
+					apiContext,
+					() => {
+						return renderPage({
+							mod,
+							renderContext,
+							env,
+							cookies: apiContext.cookies,
+						});
+					}
+				);
+			} else {
+				return await renderPage({
 					mod,
 					renderContext,
 					env,
 					cookies: apiContext.cookies,
 				});
 			}
-		);
-	} else {
-		return await renderPage({
-			mod,
-			renderContext,
-			env,
-			cookies: apiContext.cookies,
-		});
+		}
+		case 'endpoint': {
+			const result = await callEndpoint(
+				mod as any as EndpointHandler,
+				env,
+				renderContext,
+				onRequest
+			);
+			return result;
+		}
+		default:
+			throw new Error('');
```

Comment:
Would a later PR populate this error message?

Reply:
I will update the error in this PR. Totally forgot it. Thank you!